# Prepare data

## Load data

Load the original data files outputted by the bioinformatic pipeline.

### Sample metadata

```{r load_sample_metadata, warning=FALSE, comments="", message=FALSE, eval=FALSE}
sample_metadata <- read_tsv("data/DMB0157_metadata.tsv.gz") %>%
    rename(sample=1) %>%
    mutate(season = case_when(
        startsWith(sample, "EHI01") ~ "spring",
        startsWith(sample, "EHI02") ~ "autumn",
        TRUE ~ NA_character_ # Default case if no match
      )) %>%
    mutate(type = case_when(
        region %in% c("Villabona","OÃ±ati","Irun","Usurbil") ~ "natural",
        region %in% c("Eskoriatza","Aia","Errenteria","Zarautz") ~ "protected",
        region %in% c("Donostia","Orio") ~ "urban",
        TRUE ~ NA_character_ # Default case if no match
      ))%>%
  filter(sample_type=="Faecal")
```

### Read counts

```{r load_read_counts, warning=FALSE, comments="", message=FALSE, eval=FALSE}
read_counts <- read_tsv("data/DMB0157_counts.tsv.gz") %>%
  rename(genome=1) %>%
  select(c("genome",sample_metadata$sample))
```

### Genome taxonomy

```{r load_genome_metadata, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_metadata <- read_tsv("data/DMB0157_mag_info.tsv.gz") %>%
  rename(length=mag_size) %>%
  semi_join(., read_counts, by = "genome") %>% 
  arrange(match(genome,read_counts$genome))
```

### Genome base hits

```{r load_genome_hits, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_coverage <- read_tsv("data/DMB0157_coverage.tsv.gz") %>%
  rename(genome=1) %>%
  dplyr::select(c("genome",sample_metadata$sample)) %>% 
  semi_join(., genome_metadata, by = "genome")%>%
  arrange(match(genome, genome_metadata$genome))
```

### Genome tree

```{r load_genome_tree, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_tree <- read_tree("data/DMB0157.tree")
genome_tree$tip.label <- str_replace_all(genome_tree$tip.label,"'", "") #remove single quotes in MAG names
genome_tree <- keep.tip(genome_tree, tip=genome_metadata$genome) # keep only MAG tips
```

### Genome annotations

```{r load_genome_annotations, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_annotations <- read_tsv("data/genome_annotations.tsv.xz") %>%
    rename(gene=1, genome=2, contig=3)
```

## Create working objects

Transform the original data files into working objects for downstream analyses.

### Filter reads by coverage

```{r filter_coverage, warning=FALSE, comments="", message=FALSE, eval=FALSE}
min_coverage=0.3
read_counts_filt <- genome_coverage %>%
  mutate(across(where(is.numeric), ~ ifelse(. > min_coverage, 1, 0))) %>%
  mutate(across(-1, ~ . * read_counts[[cur_column()]])) 
```

### Transform reads into genome counts

```{r calculate_genome_counts_unfiltered, warning=FALSE, comments="", message=FALSE, eval=FALSE}
readlength=150
genome_counts <- read_counts %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))
```

```{r calculate_genome_counts_filtered, warning=FALSE, comments="", message=FALSE, eval=FALSE}
readlength=150
genome_counts_filt <- read_counts_filt %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))
```

****Sample without genome counts****
```{r exclut_sample, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_counts_filt %>%
  select_if(~ !is.numeric(.) || sum(.) == 0)%>%
  select(-genome) %>% 
  colnames()
```

****Mag without reads****
```{r exclut_data, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_counts_filt %>%
  filter(rowSums(across(where(is.numeric)))==0) %>% 
  column_to_rownames(., "genome") %>% 
  rownames()
```

### Filter the data

```{r filter_data, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_counts_filt <- genome_counts_filt %>%
  select_if(~ !is.numeric(.) || sum(.) != 0)%>%
  filter(rowSums(across(where(is.numeric)))!=0)

genome_metadata <- genome_metadata %>%
  semi_join(., genome_counts_filt, by = "genome")

sample_metadata <- sample_metadata %>%
      filter(sample %in% colnames(genome_counts_filt))

genome_tree <- keep.tip(genome_tree, tip=genome_metadata$genome) 
```


### Distill annotations into GIFTs 

```{r distill_annotations, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_gifts_raw <- distill(genome_annotations,GIFT_db,genomecol=2,annotcol=c(9,10,19), verbosity=F)
```
```{r distill_annotations_filt, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_counts_filt_func <- genome_counts_filt[genome_counts_filt$genome %in% rownames(genome_gifts),]
genome_gifts <- genome_gifts_raw[genome_counts_filt_func$genome,]
genome_gifts <- genome_gifts[, colSums(genome_gifts != 0) > 0]
```

## Prepare color scheme

[AlberdiLab](www.alberdilab.dk) projects use unified color schemes developed for the [Earth Hologenome Initiative](www.earthhologenome.org), to facilitate figure interpretation.

```{r get_ehi_colors, warning=FALSE, comments="", message=FALSE, eval=FALSE}
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    select(phylum, colors) %>% 
    unique() %>%
    arrange(phylum) %>%
    pull(colors, name=phylum)

location_colors <- c("#444D4F",'#3D5C61','#41B6C0','#90C8C5','#E5D388','#BFA366',"#bb6749","#834630",'#6E5244')

type_colors <- c("#e6c9e0","#bd70ae","#753568")

season_colors <- c("#009270","#ac5c4c")

```

## Prepare a phyloseq object

```{r phyloseq_objects, warning=FALSE, comments="", message=FALSE, eval=FALSE}
phylo_samples <- sample_metadata %>% 
  column_to_rownames("sample") %>% 
  sample_data() #convert to phyloseq sample_data object
phylo_genome <- genome_counts_filt %>% 
  column_to_rownames("genome") %>% 
  otu_table(., taxa_are_rows = TRUE)
phylo_taxonomy <- genome_metadata %>%
  column_to_rownames("genome") %>% 
  as.matrix() %>% 
  tax_table() #convert to phyloseq tax_table object
phylo_tree <- phy_tree(genome_tree) 

physeq_genome <- phyloseq(phylo_genome, phylo_taxonomy, phylo_samples,phylo_tree)
physeq_genome_clr <- microbiome::transform(physeq_genome, 'clr')
```

## Wrap working objects

All working objects are wrapped into a single Rdata object to facilitate downstream usage.

```{r wrap_working_objects, warning=FALSE, comments="", message=FALSE, eval=FALSE}
save(sample_metadata, 
     genome_metadata, 
     read_counts, 
     genome_counts, 
     genome_counts_filt, 
     genome_tree,
     genome_gifts_raw,
     genome_gifts, 
     phylum_colors,
     location_colors,
     type_colors,
     physeq_genome,
     physeq_genome_clr,
     file = "data/data.Rdata")
```
